### DISCUSSION TITLE
How to compile a C# program with csc built from this repo?

### DISCUSSION THREAD

**Discussion by B1Z0N** [dotnet/roslyn#60840] [Q&A]
It seems that I didn't understand something about this repo structure and/or purposes. I read the docs, built everything from source and tried  finding csc.exe or in Linux case .dll but I can't find it.  Also tried searching discussions and googling, sorry for stupidity, pls help.

I just need to compile .cs file to .dll. So that then I could run it on clr.

**Comment by CyrusNajmabadi** [comment]
I think you likely want: https://docs.microsoft.com/en-us/dotnet/core/install/linux.  Once installed you can use the `dotnet` program to build and run C#/.net apps.

**Reply by B1Z0N** [reply]
Yeah, thanks, but I also would like to make some contributions and test it right away.

**Comment by RikkiGibson** [comment]
You should be able to do this with something like the following:

1. Go to your roslyn clone and run `.\build.cmd -restore -pack` or `./build.sh --restore --pack` on Linux
2. Now open the project that you want to build using your local compiler. Add the property `<RestoreAdditionalProjectSources><PATH-TO-YOUR-ROSLYN-ENLISTMENT>/artifacts/packages/Debug/Shipping/</RestoreAdditionalProjectSources>` to the project file. (replacing the placeholder with a real path, of course)
3. Add a PackageReference to `Microsoft.Net.Compilers.Toolset` with the version you packed. This should be apparent from the Roslyn build/pack output and from the Shipping/ folder referenced in `RestoreAdditionalProjectSources`. You can add this reference via the `dotnet add myproject.csproj package Microsoft.Net.Compilers.Toolset --version 4.3.0-dev` command (replacing the placeholders and version where appropriate), or by manually adding a `<PackageReference>` to the project file.

**Note**: if you modify the compiler again and re-pack, you'll also need to find and delete the previous local compiler from your nuget cache, in order for your project to find and pick up the new one. By default the nuget cache is located at `%userprofile%\.nuget\packages` or `~/.nuget/packages` on Linux. Within this folder find `microsoft.net.compilers.toolset`, and within it find any lingering locally-packed dev compilers, e.g. a folder named `4.3.0-dev` at time of writing. Any folder suffixed with `-dev` was locally built.

The above steps are based on [this doc](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Building%2C%20Debugging%2C%20and%20Testing%20on%20Windows.md#testing-on-the-dotnetruntime-repo) which describes how to use a locally built compiler to build the runtime. We have a whole pile of msbuild logic that's specific to components of the .NET SDK, so I adapted these steps into something that should work for "normal" projects.

I hope this helps, and feel free to follow up if things don't work or if you have questions.
